ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```
@@ -0,0 +1,97 @@
+    -- action.
+  }
+
+-- | A key given after registering a clean-up action. This key can be used to
+-- 'unregister' it.
+newtype ResourceKey = ResourceKey Int
+
+-- | Register a clean-up action that will be run when the 'ResourceRegistry'
+-- is closed, unless it is 'unregister'ed before that happens.
+register :: MonadSTM m => ResourceRegistry m -> m () -> m ResourceKey
+register ResourceRegistry { _registered, _nextKey } cleanup = atomically $ do
+    key <- readTVar _nextKey
+    writeTVar _nextKey $! succ key
+    modifyTVar' _registered $ IntMap.insertWith
+      (error $ "key already registered: " <> show key)
+      key
+      cleanup
+    return $ ResourceKey key
+
+-- | Unregister a resource that already has been cleaned up. It will no longer
+-- be cleaned up when the 'ResourceRegistry' is closed.
+unregister :: MonadSTM m => ResourceRegistry m -> ResourceKey -> m ()
+unregister ResourceRegistry { _registered } (ResourceKey key) = atomically $
+    modifyTVar' _registered $ \m ->
+      assert (IntMap.member key m) $ IntMap.delete key m
+
+-- | Create new resource registry.
+new :: MonadSTM m => m (ResourceRegistry m)
+new = atomically $ do
+    _registered <- newTVar IntMap.empty
+    _nextKey    <- newTVar minBound
+    return ResourceRegistry { _registered, _nextKey }
+
+-- | Close the resource registry. All registered clean-up actions are run.
+--
+-- The exceptions thrown while calling the clean-up actions are silently
+-- discarded.
+--
+-- After closing a 'ResourceRegistry', it should no longer be used. This means
+-- that a 'ResourceRegistry' can only be closed once.
+close :: (MonadSTM m, MonadMask m) => ResourceRegistry m -> m ()
+close ResourceRegistry { _registered } = do
+    cleanups <- atomically $ do
+      cleanups <- IntMap.elems <$> readTVar _registered
+      writeTVar _registered $ error msg
+      return cleanups
+    mask $ \_unmask -> forM_ cleanups $ \cleanup ->
```

There is `mask_`.

---

Let's document that all cleanup handlers are run with exceptions masked. Depending on the resources managed, this may or may not be time consuming.

---

Let's not swallow exceptions: let's try to cleanup as much as we can, and if any of those throws an exception, pick an arbitrary (the first, say) one to rethrow -- we probably don't expect these exceptions, so if they arise, that might be indicative of a bug.